sycl/include/CL/sycl/kernel_bundle.hpp

```diff
@@ -0,0 +1,569 @@
+/////////////////////////
+
+namespace detail {
+
+bool has_kernel_bundle_impl(const context &Ctx, const std::vector<device> &Devs,
+                            bundle_state State);
+
+bool has_kernel_bundle_impl(const context &Ctx, const std::vector<device> &Devs,
+                            const std::vector<kernel_id> &kernelIds,
+                            bundle_state State);
+} // namespace detail
+
+/// Returns true if the following is true:
+/// The application defines at least one kernel that is compatible with at
+/// least one of the devices in Devs, and that kernel can be represented in a
+/// device image of state State.
+///
+/// If State is bundle_state::input, all devices in Devs have
+/// aspect::online_compiler.
+///
+/// If State is bundle_state::object, all devices in Devs have
+/// aspect::online_linker.
+template <bundle_state State>
+bool has_kernel_bundle(const context &Ctx, const std::vector<device> &Devs) {
+  return detail::has_kernel_bundle_impl(Ctx, Devs, State);
+}
+
+template <bundle_state State>
+bool has_kernel_bundle(const context &Ctx, const std::vector<device> &Devs,
+                       const std::vector<kernel_id> &KernelIDs) {
+  return detail::has_kernel_bundle_impl(Ctx, Devs, KernelIDs, State);
+}
+
+template <bundle_state State> bool has_kernel_bundle(const context &Ctx) {
+  return has_kernel_bundle<State>(Ctx, Ctx.get_devices());
+}
+
+template <bundle_state State>
+bool has_kernel_bundle(const context &Ctx,
+                       const std::vector<kernel_id> &KernelIDs) {
+  return has_kernel_bundle<State>(Ctx, Ctx.get_devices(), KernelIDs);
+}
+
+template <typename KernelName, bundle_state State>
+bool has_kernel_bundle(const context &Ctx) {
+  return has_kernel_bundle<State>(Ctx, {get_kernel_id<KernelName>()});
+}
+
+template <typename KernelName, bundle_state State>
+bool has_kernel_bundle(const context &Ctx, const std::vector<device> &Devs) {
+  return has_kernel_bundle<State>(Ctx, Devs, {get_kernel_id<KernelName>()});
+}
+
+/////////////////////////
+// is_compatible API
+/////////////////////////
+
+/// Returns true if all of the kernels identified by KernelIDs are compatible
+/// with the device Dev.
+bool is_compatible(const std::vector<kernel_id> &KernelIDs, const device &Dev);
+
+template <typename KernelName> bool is_compatible(const device &Dev) {
+  return is_compatible({get_kernel_id<KernelName>()}, Dev);
+}
+
+/////////////////////////
+// join API
+/////////////////////////
+
+namespace detail {
+
+std::shared_ptr<detail::kernel_bundle_impl>
+join_impl(const std::vector<detail::KernelBundleImplPtr> &Bundles);
+
+}
+
+/// Returns a new kernel bundle that represents the union of all the device
+/// images in the input bundles with duplicates removed.
+template <sycl::bundle_state State>
+sycl::kernel_bundle<State>
+join(const std::vector<sycl::kernel_bundle<State>> &Bundles) {
+  std::vector<detail::KernelBundleImplPtr> KernelBundleImpls;
+  // TODO: Can we just reinterpret?
```

Comment:
Could you please elaborate more? If you mean reinterpreting vector of `sycl::kernel_bundle` into vector of `sycl::detail::kernel_bundle_plain`, it'd work, since both have the same layout, but I'm not sure if it's a good idea in terms of code readability.

Reply:
Yes, to improve readability we could write a helper function which would hide the reinterpret. Additional concern here is that destructors may start working incorrectly. I will remove the TODO.